util/cloud/aws.cc

```diff
@@ -403,25 +417,56 @@ void AWS::Sign(std::string_view payload_sig, HttpHeader* header) const {
   header->set(h2::field::authorization, auth_header);
 }
 
+std::error_code AWS::SendRequest(std::string_view payload_sig, http::Client* client,
+                                 h2::request<h2::empty_body>* req,
+                                 h2::response<h2::string_body>* resp) {
+  Sign(payload_sig, req);
+
+  VLOG(1) << "Req: " << *req;
+  auto ec = client->Send(*req, resp);
+
+  // Check if session_token expired on an aws connection, established via instance metadata.
+  // In this case, try to update it, re-sign the request and re-try it.
+  if (!ec && resp->result() == h2::status::bad_request && !connection_data_.role_name.empty() &&
```

Comment:
maybe just wrap 2 conditions (bad_request and find kExpiredTokenSentinel) in an auxillary function `IsTokenRefreshNeeded` or something